### PR TITLE
Sort directory objects correctly

### PR DESCRIPTION
**Pull request by klauspost**
## Description

Decode dirobjects when listing and sort them correctly.

## How to test this PR?

@vadmeste Has reproducer.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: cmd/metacache-walk.go [hunk 1]
```diff
@@ -158,10 +166,18 @@ func (s *xlStorage) WalkDir(ctx context.Context, opts WalkDirOptions, wr io.Writ
...
+			if isDirObj {
+				meta.name = meta.name[:len(meta.name)-1] + globalDirSuffixWithSlash
+			}
+
 			meta.metadata, err = ioutil.ReadFile(pathJoin(volumeDir, meta.name, xlStorageFormatFile))
 			switch {
 			case err == nil:
 				// It was an object
+				if isDirObj {
+					meta.name = strings.TrimSuffix(meta.name, globalDirSuffixWithSlash) + slashSeparator
```
Comment (harshavardhana) [inline cmd/metacache-walk.go:179]:
```suggestion
					meta.name = decodeDirObject(meta.name)
```

File: cmd/metacache-walk.go [hunk 2]
```diff
@@ -96,9 +96,16 @@ func (s *xlStorage) WalkDir(ctx context.Context, opts WalkDirOptions, wr io.Writ
 			// Forward some errors?
 			return nil
 		}
-
+		dirObjects := make(map[string]struct{})
 		for i, entry := range entries {
 			if strings.HasSuffix(entry, slashSeparator) {
+				if strings.HasSuffix(entry, globalDirSuffixWithSlash) {
+					// Add without extension so it is sorted correctly.
+					entry = strings.TrimSuffix(entry, globalDirSuffixWithSlash) + slashSeparator
```
Comment (harshavardhana) [inline cmd/metacache-walk.go:104]:
```suggestion
					entry = decodeDirObject(entry)
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/metacache-walk.go (hunk 1 above)
Comment (klauspost) [reply cmd/metacache-walk.go:179]: No. This has a slash to sort correctly. `decodeDirObject` works only without slash.

File: cmd/metacache-walk.go (hunk 2 above)
Comment (klauspost) [reply cmd/metacache-walk.go:104]: Same

**Review by harshavardhana — APPROVED** [review]

**Comment by vadmeste** [comment]
@klauspost it looks like other calls of decodeDirObject in erasure-metadata.go and erasure-server-sets.go are not needed anymore.

**Comment by harshavardhana** [comment]
> @klauspost it looks like other calls of decodeDirObject in erasure-metadata.go and erasure-server-sets.go are not needed anymor

@vadmeste  can optimize them later

**Comment by klauspost** [comment]
@vadmeste I checked but wasn't sure.

The one in  erasure-server-sets.go is in `GetObjectInfo`, unrelated to listing.
The one in `ToObjectInfo` is used by other stuff, so I kept it.

**Review by vadmeste — APPROVED** [review]
LGTM

**Comment by harshavardhana** [comment]
> @vadmeste I checked but wasn't sure.
> 
> The one in erasure-server-sets.go is in `GetObjectInfo`, unrelated to listing.
> The one in `ToObjectInfo` is used by other stuff, so I kept it

@klauspost all those are independently needed.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#053d56fd-449b-4e6e-9b00-d84a5b9c4bb1) |
<a name="053d56fd-449b-4e6e-9b00-d84a5b9c4bb1"></a>
### 10886-0051cf6/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31040
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 14d57c97f3a6:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 53 seconds
(4/15) Running aws-sdk-ruby tests ... done in 20 seconds
(5/15) Running awscli tests ... done in 2 minutes and 47 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 3 minutes and 44 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 38 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 49 seconds
{
  "args": {
    "bucketName": "minio-go-test-6osmj0yoq5ckz4zt",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1185,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 10 minutes and 48 seconds
(10/15) Running minio-js tests ... FAILED in 46 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 80,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... done in 18 minutes and 28 seconds
(11/15) Running s3cmd tests ... done in 2 minutes and 15 seconds
(12/15) Running s3select tests ... done in 38 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by i0x71** [comment]
From what i understand, this bug was fixed in the latest release, however it seems that i am still seeing duplicate directories
i0x71@debian:~$ mc ls dot9/sys
[2020-11-11 15:07:22 EST]     0B configs/
[2020-11-12 14:21:34 EST]     0B notes/
[2020-11-12 14:50:02 EST]     0B pkg/
[2020-11-12 13:56:22 EST]     0B scripts/
[2020-11-13 18:22:17 EST]     0B configs/
[2020-11-13 18:22:17 EST]     0B notes/
[2020-11-13 18:22:17 EST]     0B pkg/
[2020-11-13 18:22:17 EST]     0B scripts/

Is there anything additional that needs to be done to make this fix effective ?
Looks like duplicates are displayed in the web gui as well.
Whats even more interesting is why the timestamp for the second set of directories shows updated at the time of minio update, directories have not been touched since yesterday.

i0x71@debian:~$ mc --version
mc version RELEASE.2020-10-03T02-54-56Z
i0x71@debian:~$ mc admin update dot9
Server `dot9` already running the most recent version 2020-11-13T20:10:18Z of MinIO

**Comment by harshavardhana** [comment]
> From what i understand, this bug was fixed in the latest release, however it seems that i am still seeing duplicate directories
> i0x71@debian:~$ mc ls dot9/sys
> [2020-11-11 15:07:22 EST] 0B configs/
> [2020-11-12 14:21:34 EST] 0B notes/
> [2020-11-12 14:50:02 EST] 0B pkg/
> [2020-11-12 13:56:22 EST] 0B scripts/
> [2020-11-13 18:22:17 EST] 0B configs/
> [2020-11-13 18:22:17 EST] 0B notes/
> [2020-11-13 18:22:17 EST] 0B pkg/
> [2020-11-13 18:22:17 EST] 0B scripts/

This issue is in `mc` @i0x71 side I have reproduced the problem.

**Comment by i0x71** [comment]
> > From what i understand, this bug was fixed in the latest release, however it seems that i am still seeing duplicate directories
> > i0x71@debian:~$ mc ls dot9/sys
> > [2020-11-11 15:07:22 EST] 0B configs/
> > [2020-11-12 14:21:34 EST] 0B notes/
> > [2020-11-12 14:50:02 EST] 0B pkg/
> > [2020-11-12 13:56:22 EST] 0B scripts/
> > [2020-11-13 18:22:17 EST] 0B configs/
> > [2020-11-13 18:22:17 EST] 0B notes/
> > [2020-11-13 18:22:17 EST] 0B pkg/
> > [2020-11-13 18:22:17 EST] 0B scripts/
> 
> This issue is in `mc` @i0x71 side I have reproduced the problem.

Thanks, i'll wait for the mc release.

**Comment by harshavardhana** [comment]
> Looks like duplicates are displayed in the web gui as well.

wait are you saying it's still visible? can you provide your directory tree structure and MinIO setup?

**Comment by harshavardhana** [comment]
I tried to create a directory structure like yours assuming you have some objects inside each of those prefixes. 

```
~  mc ls myminio/testbucket/
[2020-11-13 15:31:41 PST]     0B configs/
[2020-11-13 15:31:41 PST]     0B notes/
[2020-11-13 15:31:41 PST]     0B pkg/
[2020-11-13 15:31:41 PST]     0B scripts/
```

```
~ mc ls myminio/testbucket/ -r
[2020-11-13 15:31:00 PST]     0B configs/
[2020-11-13 15:31:20 PST]   525B configs/hosts
[2020-11-13 15:31:03 PST]     0B notes/
[2020-11-13 15:31:23 PST]   525B notes/hosts
[2020-11-13 15:31:07 PST]     0B pkg/
[2020-11-13 15:31:26 PST]   525B pkg/hosts
[2020-11-13 15:31:14 PST]     0B scripts/
[2020-11-13 15:31:29 PST]   525B scripts/hosts
```

**Comment by i0x71** [comment]
Ill get back to you with th details

**Comment by harshavardhana** [comment]
NOTE: https://github.com/minio/mc/pull/3491 is needed to make sure `mc` is not listing duplicate content

**Comment by i0x71** [comment]
It does indeed show up duplicate on the gui, as per screenshot
My directory structure is

i0x71@debian:~$ mc tree dot9/sys
dot9/sys
├─ configs
├─ notes
├─ pkg
│  └─ clusterBuilder
│     └─ payload
├─ scripts
├─ configs
├─ notes
├─ pkg
│  └─ clusterBuilder
│     └─ payload
├─ scripts
└─ yg

Start param: /opt/runtime/minio/minio server http://10.0.0.{3...9}/export/{0...1}/alpha
So 7 Hosts with 2 drives per host
Total: 14 drives online, 0 drives offline

I think this mess was created by using mc mb serv/sys/bucket as well as using mc mv serv/blah/blah serv/sys/bucket
I will try to replicate this issue and steps in a new bucket

Screenshot_2020-11-13_18-44-29

**Comment by harshavardhana** [comment]
Do you have versioning enabled on the bucket? @i0x71

**Comment by i0x71** [comment]
Nope
i0x71@debian:~$ mc version info dot9/sys
dot9/sys is un-versioned

**Comment by i0x71** [comment]
I just tried copying serv/sys/ to another bucket, and that new bucket only shows 1 copy of directories after the copy.
Is there any debug info i can profile on the duplicate existing directories ?

**Comment by harshavardhana** [comment]
```
~ mc tree sys/
sys/
├─ configs
├─ notes
├─ pkg
│  ├─ clusterBuilder
│  └─ payload
└─ yg
```

```
~ mc mirror sys/ myminio/sys/
~ mc tree myminio/sys/
myminio/sys/
├─ configs
├─ notes
├─ pkg
│  ├─ clusterBuilder
│  └─ payload
└─ yg
```

**Comment by harshavardhana** [comment]
> I just tried copying serv/sys/ to another bucket, and that new bucket only shows 1 copy of directories after the copy.
> Is there any debug info i can profile on the duplicate existing directories ?

I think perhaps this is due to some metadata cache issue, might have been triggered by some snafu that happened in the previous release due to 

- incorrect sorting
- and other prefix duplication issue

**Comment by i0x71** [comment]
Is there a way to clear this cache ?

**Comment by harshavardhana** [comment]
let's try this purging your 

>  rm -rf /tmp/disk1{3..9}/export{0..1}/.minio.sys/buckets/sys/.metacache/

**Comment by harshavardhana** [comment]
and of course updating the metacache by updating the content also should invalidate it.

**Comment by i0x71** [comment]
Actually i just listed directories again and the duplicates are gone! I haven't ran the above rm command yet.
Maybe this could have been related to the fact that i copied dot9/sys to dot9/testa ?
That or some cache expiring
Now both mc and web gui show properly

**Comment by harshavardhana** [comment]
> Actually i just listed directories again and the duplicates are gone! I haven't ran the above rm command yet.
> Maybe this could have been related to the fact that i copied dot9/sys to dot9/testa ?
> That or some cache expiring
> Now both mc and web gui show properly

yeah definitely @i0x71

**Comment by harshavardhana** [comment]
You should never see it ideally, but if you do please keep an eye out and report to us by opening an issue. Thanks @i0x71

**Comment by i0x71** [comment]
Thanks for looking into this!

**Pull request merged**